community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/AwaitIndexProcedure.java

```diff
@@ -0,0 +1,157 @@
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.neo4j.function.Predicates;
+import org.neo4j.kernel.api.KernelTransaction;
+import org.neo4j.kernel.api.ReadOperations;
+import org.neo4j.kernel.api.exceptions.ProcedureException;
+import org.neo4j.kernel.api.exceptions.Status;
+import org.neo4j.kernel.api.exceptions.index.IndexNotFoundKernelException;
+import org.neo4j.kernel.api.exceptions.schema.SchemaRuleNotFoundException;
+import org.neo4j.kernel.api.index.IndexDescriptor;
+import org.neo4j.kernel.api.index.InternalIndexState;
+
+import static java.lang.String.format;
+
+public class AwaitIndexProcedure
+{
+    private final ReadOperations operations;
+
+    public AwaitIndexProcedure( KernelTransaction tx )
+    {
+        operations = tx.acquireStatement().readOperations();
+    }
+
+    public void execute( String labelName, String propertyKeyName, long timeout, TimeUnit timeoutUnits )
+            throws ProcedureException
+    {
+        int labelId = getLabelId( labelName );
+        int propertyKeyId = getPropertyKeyId( propertyKeyName );
+        String indexDescription = formatIndex( labelName, propertyKeyName );
+        IndexDescriptor index = getIndex( labelId, propertyKeyId, indexDescription );
+        waitUntilOnline( index, indexDescription, timeout, timeoutUnits );
+    }
+
+    private int getLabelId( String labelName ) throws ProcedureException
+    {
+        int labelId = operations.labelGetForName( labelName );
+        if ( labelId == ReadOperations.NO_SUCH_LABEL )
+        {
+            throw new ProcedureException( Status.Schema.LabelAccessFailed, "No such label %s", labelName );
+        }
+        return labelId;
+    }
+
+    private int getPropertyKeyId( String propertyKeyName ) throws ProcedureException
+    {
+        int propertyKeyId = operations.propertyKeyGetForName( propertyKeyName );
+        if ( propertyKeyId == ReadOperations.NO_SUCH_PROPERTY_KEY )
+        {
+            throw new ProcedureException( Status.Schema.PropertyKeyAccessFailed,
+                    "No such property key %s", propertyKeyName );
+        }
+        return propertyKeyId;
+    }
+
+    private IndexDescriptor getIndex( int labelId, int propertyKeyId, String indexDescription ) throws
+            ProcedureException
+    {
+        try
+        {
+            return operations.indexGetForLabelAndPropertyKey( labelId, propertyKeyId );
+        }
+        catch ( SchemaRuleNotFoundException e )
+        {
+            throw new ProcedureException( Status.Schema.IndexNotFound, e, "No index on %s", indexDescription );
+        }
+    }
+
+    private void waitUntilOnline( IndexDescriptor index, String indexDescription, long timeout, TimeUnit timeoutUnits )
+            throws ProcedureException
+    {
+        try
+        {
+            Predicates.await( () -> isOnline( indexDescription, index ), timeout, timeoutUnits );
+        }
+        catch ( TunnellingException e )
```

Comment:
@davidegrohmann There was already a very half-hearted implementation of this (which did exactly what I was trying to avoid and wrapped the exception in a `RuntimeException`). I've made it more whole-hearted.